aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java

```diff
@@ -121,23 +136,49 @@ public S3FileIO(SerializableSupplier<S3Client> s3) {
    * @param s3FileIOProperties S3 FileIO properties
    */
   public S3FileIO(SerializableSupplier<S3Client> s3, S3FileIOProperties s3FileIOProperties) {
+    this(s3, null, s3FileIOProperties);
+  }
+
+  /**
+   * Constructor with custom s3 supplier, s3Async supplier and S3FileIO properties.
+   *
+   * <p>Calling {@link S3FileIO#initialize(Map)} will overwrite information set in this constructor.
+   *
+   * @param s3 s3 supplier
+   * @param s3Async s3Async supplier
+   * @param s3FileIOProperties S3 FileIO properties
+   */
+  public S3FileIO(
+      SerializableSupplier<S3Client> s3,
+      SerializableSupplier<S3AsyncClient> s3Async,
+      S3FileIOProperties s3FileIOProperties) {
     this.s3 = s3;
+    this.s3Async = s3Async;
     this.s3FileIOProperties = s3FileIOProperties;
     this.createStack = Thread.currentThread().getStackTrace();
   }
 
   @Override
   public InputFile newInputFile(String path) {
+    if (shouldUseAsyncClient()) {
+      return S3InputFile.fromLocation(path, client(), asyncClient(), s3FileIOProperties, metrics);
+    }
     return S3InputFile.fromLocation(path, client(), s3FileIOProperties, metrics);
   }
 
   @Override
   public InputFile newInputFile(String path, long length) {
+    if (shouldUseAsyncClient()) {
+      return S3InputFile.fromLocation(path, client(), asyncClient(), s3FileIOProperties, metrics);
```

Comment:
```suggestion
      return S3InputFile.fromLocation(path, length, client(), asyncClient(), s3FileIOProperties, metrics);
```
I think we miss the `length` here?